### ISSUE TITLE
routing break between 4.12.4 and 4.13.1

### ISSUE THREAD

**Issue by mykwillis** [expressjs/expressjs.com#417]
I have a route that worked properly (for me) using the 4.12.4 release, but broke when I updated to 4.13.1.

The route has an optional parameter enclosed in (literal) parenthesis, like this:

```
'/:foo\\(:bar?\\)'
```

It is intended to match paths like `/hello(world)` or `/hello()`.

On version 4.12.4, both parameters (`:foo` and `:bar`) are matched, but with version 4.13.1, it seems that only `:foo` is.

There is a gist here: https://gist.github.com/mykwillis/cf0a955d1a273257e798#file-route-js

```
C:\dev\express\route>cat route.js
/**
 * Simple Expressjs application demonstrating difference in routing behavior
 * between version 4.12.4 and 4.13.1.
 */
var express = require('express');
var request = require('request');
var http_server = express()
  .get('/:foo\\(:bar?\\)', function(req, res, next) {
    res.end(JSON.stringify({ 'foo': req.params.foo, 'bar': req.params.bar }));
  }).listen(8080, function() {
    request('http://localhost:8080/hello(world)', function(err, res, body) {
      // 4.12.4: {"foo":"hello","bar":"world"}
      // 4.13.1: {"foo":"hello"}
      console.log(body);
      process.exit();
    });
  });

C:\dev\express\route>npm install express@4.12.4
C:\dev\express\route>node route.js
{"foo":"hello","bar":"world"}

C:\dev\express\route>npm install express@4.13.1
C:\dev\express\route>node route.js
{"foo":"hello"}



```

**Comment by dougwilson** [comment]
I'll look into this, though you'll probably want to fine this kind of issue in https://github.com/strongloop/express/issues in the future :) It's possibly something from a change in `path-to-regexp`, but I'll have to check. /cc @blakeembrey if he can check it out faster than myself :)

**Comment by blakeembrey** [comment]
The matches appear to be the same for each version, I'll look into it with Express a little more.

```
➜  path-to-regexp git:(0.1.x) gco v0.1.3   
Note: checking out 'v0.1.3'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b new_branch_name

HEAD is now at 8f91a83... 0.1.3
➜  path-to-regexp git:(8f91a83) node
> require('./')('/:foo\\(:bar?\\)').exec('/hello(world)')
[ '/hello(world)',
  'hello',
  'world',
  index: 0,
  input: '/hello(world)' ]
➜  path-to-regexp git:(8f91a83) gco v0.1.6
Previous HEAD position was 8f91a83... 0.1.3
HEAD is now at 41abe34... v0.1.6
➜  path-to-regexp git:(41abe34) node
> require('./')('/:foo\\(:bar?\\)').exec('/hello(world)')
[ '/hello(world)',
  'hello',
  'world',
  index: 0,
  input: '/hello(world)' ]
```

**Comment by mykwillis** [comment]
oops, sorry about filing to the wrong repo!

**Comment by blakeembrey** [comment]
It is a regression. Likely an offset is off by one:

```
parts = []; require('./')('/:foo\\(:bar?\\)', parts); parts
[ { name: 'foo', optional: false, offset: 1 },
  { name: 0, optional: false, offset: 17 },
  { name: 'bar', optional: true, offset: 18 } ]
```

I'll patch it and make a PR now.

**Comment by dougwilson** [comment]
@blakeembrey awesome, thanks for the quick look :+1: ! I should be back from a vacay soon and get full-force back into the repos :)

**Comment by mykwillis** [comment]
thanks for the crazy fast attention, guys.

**Comment by blakeembrey** [comment]
@mykwillis I patched it in `path-to-regexp` and made a PR to `express` for you. Turned out to be a regression where it wasn't handling escaped brackets properly. Thanks for pointing it out :+1: 

Edit: See https://github.com/pillarjs/path-to-regexp/commit/22a301696755ea501154a9f466549688555303e7 for more info.

**Comment by dougwilson** [comment]
Hi @mykwillis , I'm sorry for the longer-than-usual time to patch here, but I'm back from my trip and I just released Express 4.13.2 to npm that includes the patch from @blakeembrey . Let us know if you still have any issues!

**Comment by mykwillis** [comment]
really appreciate the patch. Glad it wasn't just me going crazy with regex fever!
